template/DEVELOPING/GraphQL-frontend.md

```diff
@@ -0,0 +1,94 @@
+# GraphQL, Running Queries from React Code
+
+The React frontend uses [Apollo Client][] to dispatch GraphQL queries. What is
+nice about Apollo Client is that it understands GraphQL conventions, and stores
+response data in a normalized cache. That provides general-purpose state
+management that let's you skip writing most of the React state management code
+you would otherwise write to accompany API requests.
+
+[apollo client]: https://www.apollographql.com/docs/react/
+
+Queries are written in the GraphQL query language, but are embedded in
+TypeScript files using `gql` template tags. We have tooling in place via
+a TypeScript compiler plugin, and an ESLint plugin to check queries, and to
+provide interactive feedback in your IDE. We also have a system of code
+generation that allows TypeScript to infer accurate types for API response data,
+and for query variables.
+
+## Writing a Query
+
+GraphQL queries are written in `gql` template tags, and are dispatched using
+React hooks. Here is a basic example,
+
+```tsx
+import { gql, useQuery } from "@apollo/client"
+
+const getRecipesQuery = gql`
+  query GetRecipes {
+    recipes {
+      id
+      title
+      description
+    }
+  }
+` as import("./__generated__/get-recipes").GetRecipesDocument
+
+function RecipesList() {
+  const { data, error, loading } = useQuery(getRecipesQuery)
+  //      ^
+  //      inferred type is { recipes: { id: string, title: string, ... }[] } | undefined
+
+  if (error) {
+    return <div>Error: {error.message}</div>
+  }
+
+  if (loading) {
+    return <div>Loading...</div>
+  }
+
+  return (
+    <ul>
+      {data.recipes.map(recipe => (
+        <li>
+          {recipe.title} &mdash; {recipe.description}
+        </li>
+      ))}
+    </ul>
+  )
+}
+```
+
+See [Apollo Client][]'s documentation for usage details.
+
+> ⚠ Try to include the `id` field in every type that you request - Apollo needs
```

Comment:
I wonder if there's a way we can force/abstract that? Seems like it may introduce some challenging debugging down the line but maybe I'm just overthinking it

Reply:
I don't think there is a great way to add `id` into queries automatically. But we do have a lint in place, `graphql/required-fields`, that will error if `id` is missing; and it's smart enough to know which types do or do not have an `id` field.

Reply:
yeah that's kinda what I figured. Carry on!